### PR TITLE
Feat: Add Block User

### PR DESCRIPTION
**Pull request by Custard7**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=wh8Eb9o)
# Overview

#### 📚 What is the context and goal of this PR?

Apple App Store requires all social functionality to include the ability to block users. Plus, users should be able to block users.

#### 🥴 TL; RL:

- Adds blocking functionality to LCN

#### New LCN Plugin Invocations
```
    blockProfile: (profileId: string) => Promise<boolean>;
    unblockProfile: (profileId: string) => Promise<boolean>;
    getBlockedProfiles: () => Promise<LCNProfile[]>;
```

#### Blocking Users
      ✓ should require full auth to view blocked profiles
      ✓ allows users to view blocked profiles
      ✓ remove connection relationship after blocking a user
      ✓ remove connection requests after blocking a user 
      ✓ allows users to unblock a profile
      ✓ blocking a user should prevent receiving connection requests, VCs, VPs, and Boosts
      ✓ blocking a user should hide user from search 
      ✓ blocking a user should hide user from retrieving their profile

#### Adds new API endpoints:

- blockProfile - allows blocking another profile
- unblockProfile - allows unblocking another profile
- blocked - retrieves profiles a user has blocked

#### 🔍 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (refactor, documentation update, etc)

#### 💳 Does This Create Any New Technical Debt? ( If yes, please describe and [add JIRA TODOs](https://welibrary.atlassian.net/jira/software/projects/WE/boards/2) )
- [x] No
- [ ] Yes

# Testing

#### 🧪 Code Coverage

# Documentation

#### 📊 Storybook

# ✅ PR Checklist
- [x] Related to a Jira issue ([create one if not](https://welibrary.atlassian.net/jira/software/projects/WE/boards/2))
- [x] My code follows **style guidelines** (eslint / prettier)
- [x] I have **manually tested** common end-2-end cases
- [x] I have **reviewed** my code
- [x] I have **commented** my code, particularly where ambiguous
- [x] New and existing **unit tests pass** locally with my changes
- [ ] I have made corresponding changes to **gitbook documentation**

### 🚀 Ready to squash-and-merge?:
- [x] Code is backwards compatible
- [x] There is **not** a "Do Not Merge" label on this PR
- [x] I have thoughtfully considered the security implications of this change.
- [x] This change does not expose new public facing endpoints that do not have authentication

### CODE REVIEW

**Review by gerardopar — APPROVED** [review]
LGTM 💪🏽

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts [hunk 1]
```diff
@@ -302,3 +302,127 @@ export const getConnectionStatus = async (
     }
     return LCNProfileConnectionStatusEnum.enum.NOT_CONNECTED;
 };
+
+/** Checks whether two profiles are already connected */
```
Comment (TaylorBeeston) [inline services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:306]:
```suggestion
/** Checks if target is blocked by source */
```

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts [hunk 2]
```diff
@@ -302,3 +302,127 @@ export const getConnectionStatus = async (
...
+        (
+            await source.findRelationships({
+                alias: 'blocked',
+                where: { relationship: {}, target: { profileId: target.profileId } },
+            })
+        ).length > 0
+    );
+};
+
+/** Checks whether two profiles are already connected */
```
Comment (TaylorBeeston) [inline services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:321]:
```suggestion
/** Checks if source or target are blocking each other */
```

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts [hunk 3]
```diff
@@ -302,3 +302,127 @@ export const getConnectionStatus = async (
...
+    const [blocked, blockedBy] = await Promise.all([
+        profile.findRelationships({ alias: 'blocked' }),
+        Profile.findRelationships({
+            alias: 'blocked',
+            where: { target: { profileId: profile.profileId } },
+        }),
+    ]);
+    return [...blocked, ...blockedBy].reduce((prev, relationship): ProfileInstance[] => {
+        return [...prev, relationship.target, relationship.source];
+    }, [] as ProfileInstance[]);
```
Comment (TaylorBeeston) [inline services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:421]:
```suggestion

    return [...blocked, ...blockedBy].reduce<ProfileInstance[]>((profiles, relationship) => {
        profiles.push(relationship.target, relationship.source);

        return profiles;
    }, []);
```

nit: Reduce is one of the few places where you can get into a huge perf bottleneck from instantiating all those arrays! Likely wouldn't ever happen, but imagine someone blocks/is blocked by a million profiles!

**Review by TaylorBeeston — APPROVED** [review]
LGTM!

File: services/learn-card-network/brain-service/src/helpers/connection.helpers.ts (hunk 3 above)
Comment (Custard7) [reply services/learn-card-network/brain-service/src/helpers/connection.helpers.ts:421]: @TaylorBeeston ooh good catch, that's good to know - I remember this insight from a previous convo over in Welibrary repo now, thanks 💪 

Also thanks for the TS fix too 😎

Comment (TaylorBeeston) [reply services/learn-card-network/brain-service/src/helpers/connection.helpers.ts]: No problem! That one in particular is just burned into my brain from an article I read a while back where they solved nightmare performance issues caused by that 😂

**Pull request merged**